### PR TITLE
add module multi-head-attention

### PR DESCRIPTION
**Pull request by guillepinto**
### Description
Multi-Head Attention module used in Transformers

### Checklist
Ensure your pull request meets the following requirements:
- [x] I have reviewed existing notebooks to avoid duplications.
- [x] My notebook is fully executed.
- [x] I have included a list of all libraries and their versions used in the notebook.
- [x] The notebook file is named clearly to reflect the topic _(e.g., `cross-entropy-loss.ipynb`)._
- [x] I have placed the notebook in the appropriate directory under `notebooks/`.
- [x] My code is simple, precise, and adheres to good coding practices.

### CODE REVIEW

**Review by Factral — CHANGES_REQUESTED** [review]
very good explanations and code, the notebook is so good! (and the annexes too)

just some comments to make it even better:

- maybe will be good to name and add a code with the official implementation from pytorch of the MSA https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html , that version has implemented custom cuda kernels, so in a real architecture, it is a good practice to use it.

- the original implementation of MSA has no bias, and by default the [nn.linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) function has the bias set to true, so leave it as `bias=False` (this error is also presentend in vit.ipynb in architectures folder)

- maybe will be very good to see the attention matrix (the one with shape (len_seq,len_seq), if you have time review this: https://medium.com/@hunter-j-phillips/multi-head-attention-7924371d477a , this comment is just optional.

- I recommend rename the notebook to be multihead-self-attention.ipynb , because we are in the case of attention on the same input

**Comment by guillepinto** [comment]
> very good explanations and code, the notebook is so good! (and the annexes too)
> 
> just some comments to make it even better:
> 
> * maybe will be good to name and add a code with the official implementation from pytorch of the MSA https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html , that version has implemented custom cuda kernels, so in a real architecture, it is a good practice to use it.
> * the original implementation of MSA has no bias, and by default the [nn.linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) function has the bias set to true, so leave it as `bias=False` (this error is also presentend in vit.ipynb in architectures folder)
> * maybe will be very good to see the attention matrix (the one with shape (len_seq,len_seq), if you have time review this: https://medium.com/@hunter-j-phillips/multi-head-attention-7924371d477a , this comment is just optional.
> * I recommend rename the notebook to be multihead-self-attention.ipynb , because we are in the case of attention on the same input

- [x] pytorch implementation: not the same for me I think it may be due to differences in the implementation of internal layers, numerical precision, or the way PyTorch handles operations such as scaling, softmax, and matrix multiplication.
- [x] bias as parameter
- [x] plot of the attention probabilities in the annexes (I feel that doing so in the explanation may clutter up the explanation).
- [x] notebook rename

**Comment by Factral** [comment]
LGTM, thank u for ur attention!

**Review by Factral — APPROVED** [review]
